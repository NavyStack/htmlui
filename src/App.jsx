import 'bootstrap/dist/css/bootstrap.min.css';
import './css/Theme.css';
import './css/App.css';
import axios from 'axios';
import { React, Component } from 'react';
import { Navbar, Nav, Container } from 'react-bootstrap';
import { BrowserRouter as Router, NavLink, Redirect, Route, Switch } from 'react-router-dom';
import { Policy } from './pages/Policy';
import { Preferences } from './pages/Preferences';
import { Policies } from "./pages/Policies";
import { Repository } from "./pages/Repository";
import { Task } from './pages/Task';
import { Tasks } from './pages/Tasks';
import { Snapshots } from "./pages/Snapshots";
import { SnapshotCreate } from './pages/SnapshotCreate';
import { SnapshotDirectory } from "./pages/SnapshotDirectory";
import { SnapshotHistory } from "./pages/SnapshotHistory";
import { SnapshotRestore } from './pages/SnapshotRestore';
import { AppContext } from './contexts/AppContext';
import { UIPreferenceProvider } from './contexts/UIPreferencesContext';

export default class App extends Component {
  constructor() {
    super();

    this.state = {
      runningTaskCount: 0,
      isFetching: false,
      repoDescription: "",
    };

    this.fetchTaskSummary = this.fetchTaskSummary.bind(this);
    this.repositoryUpdated = this.repositoryUpdated.bind(this);
    this.repositoryDescriptionUpdated = this.repositoryDescriptionUpdated.bind(this);
    this.fetchInitialRepositoryDescription = this.fetchInitialRepositoryDescription.bind(this);

    const tok = document.head.querySelector('meta[name="kopia-csrf-token"]');
    if (tok && tok.content) {
      axios.defaults.headers.common['X-Kopia-Csrf-Token'] = tok.content;
    } else {
      axios.defaults.headers.common['X-Kopia-Csrf-Token'] = "-";
    }
  }

  componentDidMount() {
    const av = document.getElementById('appVersion');
    if (av) {
      // show app version after mounting the component to avoid flashing of unstyled content.
      av.style.display = "block";
    }

    this.fetchInitialRepositoryDescription();
    this.taskSummaryInterval = window.setInterval(this.fetchTaskSummary, 5000);
  }

  fetchInitialRepositoryDescription() {
    axios.get('/api/v1/repo/status').then(result => {
      if (result.data.description) {
        this.setState({
          repoDescription: result.data.description,
        });
      }
    }).catch(error => { /* ignore */ });
  }

  fetchTaskSummary() {
    if (!this.state.isFetching) {
      this.setState({ isFetching: true });
      axios.get('/api/v1/tasks-summary').then(result => {
        this.setState({ isFetching: false, runningTaskCount: result.data["RUNNING"] || 0 });
      }).catch(error => {
        this.setState({ isFetching: false, runningTaskCount: -1 });
      });
    }
  }

  componentWillUnmount() {
    window.clearInterval(this.taskSummaryInterval);
  }

  // this is invoked via AppContext whenever repository is connected, disconnected, etc.
  repositoryUpdated(isConnected) {
    if (isConnected) {
      window.location.replace("/snapshots");
    } else {
      window.location.replace("/repo");
    }
  }

  repositoryDescriptionUpdated(desc) {
    this.setState({
      repoDescription: desc,
    });
  }

  render() {
    const { uiPrefs, runningTaskCount } = this.state;

    return (
      <Router>
        <AppContext.Provider value={this}>
          <UIPreferenceProvider initalValue={uiPrefs}>
            <Navbar expand="sm" variant="light">
              <Navbar.Brand href="/"><img src="/kopia-flat.svg" className="App-logo" alt="logo" /></Navbar.Brand>
              <Navbar.Toggle aria-controls="basic-navbar-nav" />
              <Navbar.Collapse id="basic-navbar-nav">
                <Nav className="me-auto">
                  <NavLink data-testid="tab-snapshots" className="nav-link" activeClassName="active" to="/snapshots">Snapshots</NavLink>
                  <NavLink data-testid="tab-policies" className="nav-link" activeClassName="active" to="/policies">Policies</NavLink>
                  <NavLink data-testid="tab-tasks" className="nav-link" activeClassName="active" to="/tasks">Tasks <>
                    {runningTaskCount > 0 && <>({runningTaskCount})</>}
                  </>
                  </NavLink>
                  <NavLink data-testid="tab-repo" className="nav-link" activeClassName="active" to="/repo">Repository</NavLink>
                  <NavLink data-testid="tab-preferences" className="nav-link" activeClassName="active" to="/preferences">Preferences</NavLink>
                </Nav>
              </Navbar.Collapse>
            </Navbar>

            <Container fluid>
              <NavLink to="/repo" style={{ color: "inherit", textDecoration: "inherit" }}>
                <h5 className="mb-4">{this.state.repoDescription}</h5>
              </NavLink>

              <Switch>
                <Route path="/snapshots/new" component={SnapshotCreate} />
                <Route path="/snapshots/single-source/" component={SnapshotHistory} />
                <Route path="/snapshots/dir/:oid/restore" component={SnapshotRestore} />
                <Route path="/snapshots/dir/:oid" component={SnapshotDirectory} />
                <Route path="/snapshots" component={Snapshots} />
                <Route path="/policies/edit/" component={Policy} />
                <Route path="/policies" component={Policies} />
                <Route path="/tasks/:tid" component={Task} />
                <Route path="/tasks" component={Tasks} />
                <Route path="/repo" component={Repository} />
                <Route path="/preferences" component={Preferences} />
                <Route exact path="/">
                  <Redirect to="/snapshots" />
                </Route>
              </Switch>
            </Container>
          </UIPreferenceProvider>
        </AppContext.Provider>
      </Router>
    );
  }
}
