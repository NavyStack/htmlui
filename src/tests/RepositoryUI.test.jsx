import axios from 'axios';
import React from 'react';
import App from '../App';
import { render } from '@testing-library/react';

jest.mock('axios');

// Wrapper
let wrapper;

/**
 * 
 */
beforeEach(() => {
    wrapper = render(<App />)
});

describe('Testing the app', () => {
    test('Should render the app', () => {
        expect(wrapper).toMatchSnapshot();
    })
})
