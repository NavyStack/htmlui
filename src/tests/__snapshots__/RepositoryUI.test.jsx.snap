// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the app Should render the app 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-sm navbar-light"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          <img
            alt="logo"
            class="App-logo"
            src="/kopia-flat.svg"
          />
        </a>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              aria-current="page"
              class="nav-link active"
              data-testid="tab-snapshots"
              href="/snapshots"
            >
              Snapshots
            </a>
            <a
              class="nav-link"
              data-testid="tab-policies"
              href="/policies"
            >
              Policies
            </a>
            <a
              class="nav-link"
              data-testid="tab-tasks"
              href="/tasks"
            >
              Tasks 
            </a>
            <a
              class="nav-link"
              data-testid="tab-repo"
              href="/repo"
            >
              Repository
            </a>
            <a
              class="nav-link"
              data-testid="tab-preferences"
              href="/preferences"
            >
              Preferences
            </a>
          </div>
        </div>
      </nav>
      <div
        class="container-fluid"
      >
        <a
          href="/repo"
          style="text-decoration: inherit;"
        >
          <h5
            class="mb-4"
          />
        </a>
        <div
          class="spinner-border text-primary"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-sm navbar-light"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        <img
          alt="logo"
          class="App-logo"
          src="/kopia-flat.svg"
        />
      </a>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="me-auto navbar-nav"
        >
          <a
            aria-current="page"
            class="nav-link active"
            data-testid="tab-snapshots"
            href="/snapshots"
          >
            Snapshots
          </a>
          <a
            class="nav-link"
            data-testid="tab-policies"
            href="/policies"
          >
            Policies
          </a>
          <a
            class="nav-link"
            data-testid="tab-tasks"
            href="/tasks"
          >
            Tasks 
          </a>
          <a
            class="nav-link"
            data-testid="tab-repo"
            href="/repo"
          >
            Repository
          </a>
          <a
            class="nav-link"
            data-testid="tab-preferences"
            href="/preferences"
          >
            Preferences
          </a>
        </div>
      </div>
    </nav>
    <div
      class="container-fluid"
    >
      <a
        href="/repo"
        style="text-decoration: inherit;"
      >
        <h5
          class="mb-4"
        />
      </a>
      <div
        class="spinner-border text-primary"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
